src/std_extensions/quantum.rs

```diff
@@ -1,19 +1,190 @@
 //! Basic HUGR quantum operations
 
+use std::cmp::max;
+use std::f64::consts::TAU;
+use std::num::NonZeroU64;
+
 use smol_str::SmolStr;
 
-use crate::extension::prelude::{BOOL_T, QB_T};
+use crate::extension::prelude::{BOOL_T, ERROR_TYPE, QB_T};
 use crate::extension::{ExtensionId, SignatureError};
 use crate::std_extensions::arithmetic::float_types::FLOAT64_TYPE;
 use crate::type_row;
-use crate::types::type_param::TypeArg;
-use crate::types::FunctionType;
+use crate::types::type_param::{TypeArg, TypeArgError, TypeParam};
+use crate::types::{ConstTypeError, CustomCheckFailure, CustomType, FunctionType, Type, TypeBound};
+use crate::utils::collect_array;
+use crate::values::CustomConst;
 use crate::Extension;
 
 use lazy_static::lazy_static;
 
 /// The extension identifier.
 pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("quantum");
+
+/// Identifier for the angle type.
+const ANGLE_TYPE_ID: SmolStr = SmolStr::new_inline("angle");
+
+fn angle_custom_type(log_denom_arg: TypeArg) -> CustomType {
+    CustomType::new(ANGLE_TYPE_ID, [log_denom_arg], EXTENSION_ID, TypeBound::Eq)
+}
+
+/// Angle type with a given log-denominator (specified by the TypeArg).
+///
+/// This type is capable of representing angles that are multiples of 2π / 2^N where N is the
+/// log-denominator.
+pub(super) fn angle_type(log_denom_arg: TypeArg) -> Type {
+    Type::new_extension(angle_custom_type(log_denom_arg))
+}
+
+/// The smallest forbidden log-denominator.
+pub const LOG_DENOM_BOUND: u8 = 54;
```

Comment:
super-nit: I find `LOG_DENOM_MAX` clearer (without having to read the doc) as to whether it's inclusive/exclusive; then you'd change `<` to `<=`

Reply:
Changed.